File: Plugins/Kingbotk/AWB Plugin/WPAlbums.vb
Friend NotInheritable Class WPAlbums
    Inherits PluginBase

    Friend Sub New()
        MyBase.New("Albums") ' Specify alternate names only

        Dim params(-1) As TemplateParameters

        OurSettingsControl = New GenericWithWorkgroups("WikiProject Albums", "Album", True, False, params)
    End Sub

    ' Settings:
    Private OurTab As New TabPage("Albums")
    Private WithEvents OurSettingsControl As GenericWithWorkgroups
    Private Const conEnabled As String = "AlbumEnabled"
    Protected Friend Overrides ReadOnly Property PluginShortName() As String
        Get
            Return "Albums"
        End Get
    End Property
    Protected Overrides ReadOnly Property PreferredTemplateName() As String
        Get
            Return "Album"
        End Get
    End Property

    Protected Overrides Sub ImportanceParameter(ByVal Importance As Importance)
        Template.NewOrReplaceTemplateParm("importance", Importance.ToString, Me.Article, False, False)
    End Sub
    Protected Friend Overrides ReadOnly Property GenericSettings() As IGenericSettings
        Get
            Return OurSettingsControl
        End Get
    End Property
    Protected Overrides ReadOnly Property CategoryTalkClassParm() As String
        Get
            Return "NA"
        End Get
    End Property
    Protected Overrides ReadOnly Property TemplateTalkClassParm() As String
        Get
            Return "NA"
        End Get
    End Property
    'Protected Overrides ReadOnly Property PreferredTemplateNameRegexString() As String
    '    Get
    '        Return "^[Aa]lbum$"
    '    End Get
    'End Property

    ' Initialisation:
    Protected Friend Overrides Sub Initialise()
        OurMenuItem = New ToolStripMenuItem("Albums Plugin")
        MyBase.InitialiseBase() ' must set menu item object first
        OurTab.UseVisualStyleBackColor = True
        OurTab.Controls.Add(OurSettingsControl)
    End Sub

    ' Article processing:
    Protected Overrides Function SkipIfContains() As Boolean
        Return False
    End Function
    Protected Overrides Sub ProcessArticleFinish()
        StubClass()
    End Sub
    Protected Overrides Function TemplateFound() As Boolean
        ' Nothing to do here
    End Function
    Protected Overrides Sub GotTemplateNotPreferredName(ByVal TemplateName As String)
        ' Only WPBio used to do something here (if {{musician}} add to musician-work-group)
    End Sub
    Protected Overrides Function WriteTemplateHeader(ByRef PutTemplateAtTop As Boolean) As String
        WriteTemplateHeader = "{{album" & WriteOutParameterToHeader("class") & _
           WriteOutParameterToHeader("importance")
    End Function

    'User interface:
    Protected Overrides Sub ShowHideOurObjects(ByVal Visible As Boolean)
        PluginManager.ShowHidePluginTab(OurTab, Visible)
    End Sub

    ' XML settings:
    Protected Friend Overrides Sub ReadXML(ByVal Reader As System.Xml.XmlTextReader)
        Dim blnNewVal As Boolean = PluginManager.XMLReadBoolean(Reader, conEnabled, Enabled)
        If Not blnNewVal = Enabled Then Enabled = blnNewVal ' Mustn't set if the same or we get extra tabs
        OurSettingsControl.ReadXML(Reader)
    End Sub
    Protected Friend Overrides Sub Reset()
        OurSettingsControl.Reset()
    End Sub
    Protected Friend Overrides Sub WriteXML(ByVal Writer As System.Xml.XmlTextWriter)
        Writer.WriteAttributeString(conEnabled, Enabled.ToString)
        OurSettingsControl.WriteXML(Writer)
    End Sub

    ' Misc:
    Protected Overrides ReadOnly Property InspectUnsetParameters() As Boolean
        Get
            Return False
        End Get
    End Property
    Protected Overrides Sub InspectUnsetParameter(ByVal Param As String)
    End Sub ' will never be called
    Friend Overrides ReadOnly Property HasReqPhotoParam() As Boolean
        Get
            Return False
        End Get
    End Property
    Friend Overrides Sub ReqPhoto()
    End Sub
End Class

File: AWB/Plugins/Kingbotk Plugin/WikiProjects/WPAlbums.cs
using AutoWikiBrowser.Plugins.Kingbotk;
using AutoWikiBrowser.Plugins.Kingbotk.Plugins;
using System.Windows.Forms;
using System.Xml;

internal sealed class WPAlbums : PluginBase
{
    private const string PluginName = "WikiProject Albums";

    private const string Prefix = "Albums";

    internal WPAlbums() : base("Album|Albums|WP Albums|WPAlbums")
    {
        // Specify alternate names only

        TemplateParameters[] @params = new TemplateParameters[-1 + 1];

        _ourSettingsControl = new GenericWithWorkgroups("WikiProject Albums", Prefix, true, @params);
    }

    // Settings:
    private readonly TabPage _ourTab = new TabPage("Albums");

    private readonly GenericWithWorkgroups _ourSettingsControl;

    protected internal override string PluginShortName
    {
        get { return "Albums"; }
    }

    protected override string PreferredTemplateName
    {
        get { return PluginName; }
    }

    protected override void ImportanceParameter(Importance importance)
    {
        Template.NewOrReplaceTemplateParm("importance", importance.ToString(), TheArticle, false, false);
    }

    protected internal override IGenericSettings GenericSettings
    {
        get { return _ourSettingsControl; }
    }

    // Initialisation:
    protected internal override void Initialise()
    {
        OurMenuItem = new ToolStripMenuItem("Albums Plugin");
        InitialiseBase();
        // must set menu item object first
        _ourTab.UseVisualStyleBackColor = true;
        _ourTab.Controls.Add(_ourSettingsControl);
    }

    // Article processing:
    protected override bool SkipIfContains()
    {
        return false;
    }

    protected override void ProcessArticleFinish()
    {
        StubClass();
    }

    protected override bool TemplateFound()
    {
        // Nothing to do here
        return false;
    }

    protected override string WriteTemplateHeader()
    {
        return "{{album" + WriteOutParameterToHeader("class") + WriteOutParameterToHeader("importance");
    }

    //User interface:
    protected override void ShowHideOurObjects(bool visible)
    {
        PluginManager.ShowHidePluginTab(_ourTab, visible);
    }

    // XML settings:
    protected internal override void ReadXML(XmlTextReader reader)
    {
        Enabled = PluginManager.XMLReadBoolean(reader, Prefix + "Enabled", Enabled);
        _ourSettingsControl.ReadXML(reader);
    }

    protected internal override void Reset()
    {
        _ourSettingsControl.Reset();
    }

    protected internal override void WriteXML(XmlTextWriter writer)
    {
        writer.WriteAttributeString(Prefix + "Enabled", Enabled.ToString());
        _ourSettingsControl.WriteXML(writer);
    }

    // Misc:
    internal override bool HasReqPhotoParam
    {
        get { return false; }
    }

    internal override void ReqPhoto()
    {
    }
}